llvm/lib/Transforms/Utils/InlineFunction.cpp

```diff
@@ -1389,6 +1389,12 @@ static void AddParamAndFnBasicAttributes(const CallBase &CB,
         if (!Arg)
           continue;
 
+        if (AL.hasParamAttr(I, Attribute::ByVal))
+          // It's unsound to propagate memory attributes to byval arguments.
+          // Even if CalledFunction doesn't e.g. write to the argument,
+          // the call to NewInnerCB may write to its by-value copy.
+          continue;
+
```

Comment:
I am slightly partial to implementing this by dropping memory attrs if we have `ByVal` at the end (similiar to how we `readnone` + `readonly`/`writeonly` conflicts). I have plans to expand this function and think it will be easier to scale we handle conflicts that way.

Reply:
This would work too, but would also drop `readonly` etc. from calls which already had `readonly byval(...)` before inlining - not sure how much this matters though.

Reply:
You could keep the original attributes and only drop if they are new?

Reply:
I think skipping byval in generaly makes sense. E.g. we also wouldn't want to propagate align or dereferenceable to it.

Reply:
I guess I feel like we will inevitably hit a case where we **do** want to propagate an attr to byval, and that point, this code will become alot more complex. Although I guess the other hand is having the list out each attr to cross out with byval is also fairly complex.

Reply:
No gripe with this going in as is.